File: src/lib/api/tts.ts
/**
 * TTS API
 *
 * Tauri commands for native text-to-speech.
 */

import { invoke } from '@tauri-apps/api/core';
import { listen, type UnlistenFn } from '@tauri-apps/api/event';
import type { VoiceInfo, TtsInitResponse, TtsState } from '../schemas';

// Commands

export async function ttsInit(): Promise<TtsInitResponse> {
  return invoke('tts_init');
}

export async function ttsListVoices(): Promise<{ voices: VoiceInfo[] }> {
  return invoke('tts_list_voices');
}

export async function ttsSpeak(
  text: string,
  interrupt = false
): Promise<{ chunkId: string }> {
  return invoke('tts_speak', { text, interrupt });
}

export async function ttsSpeakLong(
  text: string,
  interrupt = false
): Promise<{ totalChunks: number }> {
  return invoke('tts_speak_long', { text, interrupt });
}

export async function ttsStop(): Promise<{ success: boolean }> {
  return invoke('tts_stop');
}

export async function ttsPause(): Promise<{ success: boolean; supported: boolean }> {
  return invoke('tts_pause');
}

export async function ttsResume(): Promise<{ success: boolean }> {
  return invoke('tts_resume');
}

export async function ttsSetVoice(
  voiceId: string
): Promise<{ success: boolean; voice: VoiceInfo }> {
  return invoke('tts_set_voice', { voiceId });
}

export async function ttsSetRate(rate: number): Promise<{ success: boolean; rate: number }> {
  return invoke('tts_set_rate', { rate });
}

export async function ttsGetState(): Promise<TtsState> {
  return invoke('tts_get_state');
}

export async function ttsCheckCapabilities(): Promise<{
  supportsPause: boolean;
  supportsResume: boolean;
  supportsPitch: boolean;
  supportsVolume: boolean;
  supportsRate: boolean;
  minRate: number;
  maxRate: number;
}> {
  return invoke('tts_check_capabilities');
}

// Event Types

export interface TtsChunkStartedEvent {
  chunkIndex: number;
  totalChunks: number;
  text: string;
}

export interface TtsChunkCompletedEvent {
  chunkIndex: number;
}

export interface TtsCompletedEvent {
  success: boolean;
}

// Event Listeners

export function onTtsChunkStarted(
  callback: (event: TtsChunkStartedEvent) => void
): Promise<UnlistenFn> {
  return listen<TtsChunkStartedEvent>('tts:chunk-started', (event) =>
    callback(event.payload)
  );
}

export function onTtsChunkCompleted(
  callback: (event: TtsChunkCompletedEvent) => void
): Promise<UnlistenFn> {
  return listen<TtsChunkCompletedEvent>('tts:chunk-completed', (event) =>
    callback(event.payload)
  );
}

export function onTtsCompleted(
  callback: (event: TtsCompletedEvent) => void
): Promise<UnlistenFn> {
  return listen<TtsCompletedEvent>('tts:completed', (event) => callback(event.payload));
}

File: src-tauri/src/commands/tts.rs
use crate::db::models::{TtsInitResponse, VoiceInfo};
use crate::tts::{TtsCapabilities, TtsEngine, TtsStateResponse};
use serde::{Deserialize, Serialize};
use tauri::{AppHandle, State};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeakResponse {
    pub chunk_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StopResponse {
    pub success: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PauseResponse {
    pub success: bool,
    pub supported: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResumeResponse {
    pub success: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetVoiceResponse {
    pub success: bool,
    pub voice: VoiceInfo,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SetRateResponse {
    pub success: bool,
    pub rate: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListVoicesResponse {
    pub voices: Vec<VoiceInfo>,
}

/// Initialize the TTS engine
#[tauri::command]
#[specta::specta]
pub fn tts_init(state: State<'_, TtsEngine>) -> Result<TtsInitResponse, String> {
    state.init()
}

/// List available voices
#[tauri::command]
#[specta::specta]
pub fn tts_list_voices(state: State<'_, TtsEngine>) -> Result<ListVoicesResponse, String> {
    let voices = state.list_voices()?;
    Ok(ListVoicesResponse { voices })
}

/// Speak text
#[tauri::command]
#[specta::specta]
pub fn tts_speak(
    state: State<'_, TtsEngine>,
    text: String,
    interrupt: Option<bool>,
) -> Result<SpeakResponse, String> {
    let chunk_id = state.speak(&text, interrupt.unwrap_or(false))?;
    Ok(SpeakResponse { chunk_id })
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpeakLongResponse {
    pub total_chunks: usize,
}

/// Speak long text by splitting into chunks with progress events
#[tauri::command]
#[specta::specta]
pub fn tts_speak_long(
    app: AppHandle,
    state: State<'_, TtsEngine>,
    text: String,
    interrupt: Option<bool>,
) -> Result<SpeakLongResponse, String> {
    let total_chunks = state.speak_long(app, text, interrupt.unwrap_or(false))?;
    Ok(SpeakLongResponse { total_chunks })
}

/// Stop speaking
#[tauri::command]
#[specta::specta]
pub fn tts_stop(state: State<'_, TtsEngine>) -> Result<StopResponse, String> {
    let success = state.stop()?;
    Ok(StopResponse { success })
}

/// Pause speaking
#[tauri::command]
#[specta::specta]
pub fn tts_pause(state: State<'_, TtsEngine>) -> Result<PauseResponse, String> {
    let (success, supported) = state.pause()?;
    Ok(PauseResponse { success, supported })
}

/// Resume speaking
#[tauri::command]
#[specta::specta]
pub fn tts_resume(state: State<'_, TtsEngine>) -> Result<ResumeResponse, String> {
    let success = state.resume()?;
    Ok(ResumeResponse { success })
}

/// Set voice
#[tauri::command]
#[specta::specta]
pub fn tts_set_voice(
    state: State<'_, TtsEngine>,
    voice_id: String,
) -> Result<SetVoiceResponse, String> {
    let voice = state.set_voice(&voice_id)?;
    Ok(SetVoiceResponse {
        success: true,
        voice,
    })
}

/// Set speech rate
#[tauri::command]
#[specta::specta]
pub fn tts_set_rate(state: State<'_, TtsEngine>, rate: f64) -> Result<SetRateResponse, String> {
    let rate = state.set_rate(rate)?;
    Ok(SetRateResponse {
        success: true,
        rate,
    })
}

/// Get current TTS state
#[tauri::command]
#[specta::specta]
pub fn tts_get_state(state: State<'_, TtsEngine>) -> Result<TtsStateResponse, String> {
    state.get_state()
}

/// Check TTS capabilities
#[tauri::command]
#[specta::specta]
pub fn tts_check_capabilities(state: State<'_, TtsEngine>) -> Result<TtsCapabilities, String> {
    state.get_capabilities()
}
